### PR TITLE
Use a pool instead of a single connection, and reorganize the worker around that

### LINKED ISSUE #148: Procrastinate should be using a pool
It just occured to me that we're not using an aiopg pool, but a connection directly. In the defer side, there may be hundreds of concurrent coroutines that try to defer at the same time. As of today, I think they will crash. We need a pool instead (which could very well be of size 1 if people don't want to create too many connections). Having a pool means aiopg manages the "semaphore" for us and we don't have to deal with it.

Who's in for a pool party ?
@elemoine / @k4nar: I believe this will be blocking for you quickly, so let's address that!

Also, we need a parameter to configure the pool size.
Finally, I'd advise against doing anything not based on top of #147.
(note: I have the beginning of a PR, but nothing that cannot be redone)

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #148
Closes #166 

Sorry 100000000 times @elemoine 

### Successful PR Checklist:
- [ ] Tests
- [ ] Documentation (optionally: [run spell checking](https://github.com/peopledoc/procrastinate/blob/master/CONTRIBUTING.rst#build-the-documentation))
- [ ] Had a good time contributing? (if not, feel free to give some feedback)

### CODE REVIEW

File: tests/unit/test_signals.py [hunk 1]
```diff
@@ -32,3 +33,31 @@ def stop():
...
+
+def test_on_stop_signal_twice(kill_own_pid):
+    with pytest.raises(KeyboardInterrupt):
+        with signals.on_stop(callback=lambda: None):
+            kill_own_pid(signal=signal.SIGINT)
+            kill_own_pid(signal=signal.SIGINT)
+
+
+@pytest.mark.asyncio
+def test_on_stop_work_with_asyncio(kill_own_pid):
```
Comment (ewjoachim) [inline tests/unit/test_signals.py]: await

File: procrastinate/app.py [hunk 2]
```diff
@@ -203,32 +217,23 @@ def perform_import_paths(self):
...
+        wait: bool = True,
     ) -> None:
         """
-        Run a worker. This worker will run in the foreground
-        and the function will not return until the worker stops
-        (most probably when it receives a stop signal) (except if
-        `only_once` is True).
+        Run a worker. This worker will run in the foreground and the function will not
+        return until the worker stops (most probably when it receives a stop signal)
+        (except if `only_once` is True).
```
Comment (ewjoachim) [inline procrastinate/app.py]: to fix

File: procrastinate/aiopg_connector.py [hunk 3]
```diff
@@ -120,30 +166,35 @@ def make_dynamic_query(self, query: str, **identifiers: str) -> str:
...
+                )
+
+            # We'll leave this loop with a CancelError, when we get cancelled
+            while True:
+                event.set()
+                # TODO: because of https://github.com/aio-libs/aiopg/issues/249,
+                # we could get stuck in here forever if the connection closes.
+                # Maybe we should set a timeout and if connection is closed, reopen
+                # a new one.
+                await connection.notifies.get()
```
Comment (ewjoachim) [inline procrastinate/aiopg_connector.py]: (why did we change the order of event.set() and notifies.get() ? I can't remember)

File: procrastinate/aiopg_connector.py [hunk 4]
```diff
@@ -120,30 +166,35 @@ def make_dynamic_query(self, query: str, **identifiers: str) -> str:
...
+        async with self._pool.acquire() as connection:
+            for channel_name in channels:
+                await self._execute_query_connection(
+                    connection=connection,
+                    query=self.make_dynamic_query(
+                        query=sql.queries["listen_queue"], channel_name=channel_name
+                    ),
+                )
+
+            # We'll leave this loop with a CancelError, when we get cancelled
```
Comment (ewjoachim) [inline procrastinate/aiopg_connector.py]:
```suggestion
            # We'll leave this loop with a CancelledError, when we get cancelled
```

File: procrastinate/app.py [hunk 5]
```diff
@@ -57,6 +60,14 @@ def __init__(
             A :py:func:`App.task` that has a custom "name" parameter, that is not
             imported and whose module path is not in this list will
             fail to run.
+        worker_timeout:
+            This parameter indicates the maximum duration (in seconds) procrastinate
+            workers wait between each database job pull. Job activity will be pushed
+            from the db to the worker, but in case the push mechanism fails somehow,
+            workers will not stay idle longer than the number of seconds indicated by
+            this parameters.
```
Comment (ewjoachim) [inline procrastinate/app.py]:
```suggestion
            this parameter.
```

File: procrastinate/signals.py [hunk 6]
```diff
@@ -1,17 +1,29 @@
...
+# This all mean we have to save the previous signals before installing ours, and for
+# uninstalling, we need to remove the async handler and only then re-add the normal
+# one. And hope that there was no previously set async handler.
 
 
 @contextmanager
-def on_stop(callback: Callable[[signal.Signals, FrameType], None]):
+def on_stop(callback: Callable[[], None]):
     if threading.current_thread() is not threading.main_thread():
         logger.debug(
```
Comment (ewjoachim) [inline procrastinate/signals.py]: warning

File: procrastinate/signals.py [hunk 7]
```diff
@@ -22,19 +34,29 @@ def on_stop(callback: Callable[[signal.Signals, FrameType], None]):
...
         yield
     finally:
         logger.debug(
             "Resetting previous signal handler",
             extra={"action": "uninstall_signal_handlers"},
         )
+        loop.remove_signal_handler(signal.SIGINT)
+        loop.remove_signal_handler(signal.SIGTERM)
         signal.signal(signal.SIGINT, sigint_handler)
         signal.signal(signal.SIGTERM, sigterm_handler)
```
Comment (ewjoachim) [inline procrastinate/signals.py]: These 4 lines should be in `else`

File: procrastinate/utils.py [hunk 8]
```diff
@@ -67,21 +67,35 @@ def wrap_one(cls: Type, attribute_name: str):
...
     # Keep only async def methods
-    if not asyncio.iscoroutinefunction(attribute):
+    if not asyncio.iscoroutinefunction(wrapped):
         return
 
     # Create a wrapper that will call the method in a run_until_complete
-    @functools.wraps(attribute)
-    def wrapper(self, *args, **kwargs):
-        awaitable = attribute(self, *args, **kwargs)
+    # @functools.wraps(attribute)
```
Comment (ewjoachim) [inline procrastinate/utils.py]:
```suggestion
    @functools.wraps(attribute)
```

File: tests/unit/test_signals.py [hunk 9]
```diff
@@ -32,3 +33,31 @@ def stop():
...
+
+    def stop():
+        event.set()
+
+    async def wait_and_kill():
+        await asyncio.sleep(0.001)
+        kill_own_pid()
+
+    try:
+        with signals.on_stop():
```
Comment (ewjoachim) [inline tests/unit/test_signals.py]:
```suggestion
        with signals.on_stop(stop):
```

File: procrastinate/app.py (hunk 2 above)
Comment (ewjoachim) [inline procrastinate/app.py]:
```suggestion
        return until the worker stops (most probably when it receives a stop signal).
```

**Comment by ewjoachim** [comment]
Ok, I added a bunch of commits to finish fixing tests and because there was a lot missing for coverage.

I think we're good now, @elemoine you can re-review if you want.

**Comment by ewjoachim** [comment]
New implementation of signals fails with Py3.6 only 😭

**Comment by elemoine** [comment]
> New implementation of signals fails with Py3.6 only sob

The issue is that `asyncio.get_event_loop()` does not return `None` if not running in the context of an asyncio task. Instead, it creates a event loop and returns it.

See the commit that I added to the branch for a fix. The fix involves using `asyncio.Task.current_task()` on Python 3.6 to test whether running in task context or not.

**Comment by elemoine** [comment]
Might be worth cleaning up the Git commit history before merging. I let you decide @ewjoachim.

**Comment by ewjoachim** [comment]
> Might be worth cleaning up the Git commit history before merging.

I could but I think the risk of screwing up is higher than the gain. Let the commits tell the story ! :D

**Pull request merged**